### PR TITLE
fix: blocking client nbs

### PR DESCRIPTION
**Pull request by ryanswrt**
Contains #72 

This pr adds a blocking provider, allowing users to write code without calling await everywhere and using an asycio runtime.

It also adds and implements support for the RUSH_RESTORE_BY_DEFAULT env var, allowing us to easily test restored + unrestored workflows in CI (and cleans up the broken hideline stuff we tried in the quickstarts)

It also makes type inference a bit better in pyright by not relying on exec.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
 
-
 <!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
 
 # Quickstart
 
-This document will walk through executing jobs on the Rush platform. For
-a comprehensive guide on the concepts and constructing a full workflow,
-see the [full rush-py
+This document will walk through executing jobs on the Rush platform, by
```
Comment (seankhl) [inline README.md:8]:
```suggestion
This document will walk through executing jobs on the Rush platform by
```

File: rush/legacy_types.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import sys
+from typing import Any, Union
```
Comment (seankhl) [inline rush/legacy_types.py]: We should conditionally import `Union` too

File: flake.nix [hunk 3]
```diff
@@ -2,7 +2,7 @@
   description = "Rush Python SDK";
 
   inputs = {
-    nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.11";
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
```
Comment (seankhl) [inline flake.nix:5]: Should we only use unstable for 3.12?

File: flake.nix [hunk 4]
```diff
@@ -14,23 +14,38 @@
...
           projectDir = ./.;
+          python = pkgs.python312;
+          overrides = inputs.qdx-python-flake-parts.lib.withPoetryOverrides
+            (self: super: { pip = pkgs.python312Packages.pip; });
+          extraPackages = [ pkgs.quarto ];
+          blackCheck = {
+            extraCmd =
+              "black ./nbs --config pyproject.ipynb.toml --ipynb --check";
+          };
+        };
```
Comment (seankhl) [inline flake.nix:25]: I feel like 3.11 should be the default and 3.12 can be its own named target.

**Review by seankhl — CHANGES_REQUESTED** [review]
- We should be using `download().read_text()`, not `download().open().read()`, since the formaer automatically closes the file after reading and is simpler and clearer.
  - `(conformer,) = ` should be rewritten as `conformer, = ` without parens. Parens should be omitted for any destructuring assignments.

There are a few other notebook changes, I'm going to go through them now, so another review will come.

File: flake.nix (hunk 4 above)
Comment (seankhl) [reply flake.nix:25]: Hmm though running the notebooks on 3.12 will result in nicer types in the output.

File: rush/provider.py [hunk 5]
```diff
@@ -1337,6 +1214,193 @@ async def poll_module_instance(
...
+            provider: "BaseProvider | None",
+            id: UUID | None = None,
+            source: UUID | None = None,
+            value: T | None = None,
+            typeinfo: dict[str, Any] | SCALARS | None = None,
+        ):
+            super().__init__(provider, id, source, value, typeinfo)
+
+        def get(self) -> T:
+            print("Blocking get")
```
Comment (seankhl) [inline rush/provider.py]: Delete this line?

File: flake.nix (hunk 4 above)
Comment (ryanswrt) [reply flake.nix:25]: Yep that's what I thought

File: flake.nix (hunk 3 above)
Comment (ryanswrt) [reply flake.nix:5]: Not possible at the moment to have 2 different versions of nixpkgs passed to poetry

**Review by seankhl — DISMISSED** [review]
- index.py section 2.2: do we want the text "Blocking get" to appear in the output?
  - Likewise with `read_text()` we should use `write_text()` too instead of printing to an opened file or other such things
  - See the latest commit for more expected changes that should be applied similarly to the index, explainer, and tutorials

**Comment by seankhl** [comment]
One last thing: it would be good to homogeneize all the tag strings, like "{nb_name}_quickstart". Right now they're not super uniform.

**Review by seankhl — APPROVED** [review]

**Pull request merged**